src/Storages/ReadInOrderOptimizer.cpp

```diff
@@ -50,65 +53,98 @@ InputOrderInfoPtr ReadInOrderOptimizer::getInputOrder(const StorageMetadataPtr &
+                for (const auto & action : actions)
                 {
-                    current_direction = 0;
-                    break;
+                    if (action.node->type != ActionsDAG::ActionType::FUNCTION)
+                    {
+                        continue;
+                    }
+
+                    if (found_function)
+                    {
+                        current_direction = 0;
+                        break;
+                    }
+                    else
+                        found_function = true;
+
+                    if (action.node->children.size() != 1 || action.node->children.at(0)->result_name != sorting_key_columns[i])
+                    {
+                        current_direction = 0;
+                        break;
+                    }
+
+                    const auto & func = *action.node->function_base;
+                    if (!func.hasInformationAboutMonotonicity())
+                    {
+                        current_direction = 0;
+                        break;
+                    }
+
+                    auto monotonicity = func.getMonotonicityForRange(*func.getArgumentTypes().at(0), {}, {});
+                    if (!monotonicity.is_monotonic)
+                    {
+                        current_direction = 0;
+                        break;
+                    }
+                    else if (!monotonicity.is_positive)
+                        current_direction *= -1;
                 }
-                else
-                    found_function = true;
 
-                if (action.node->children.size() != 1 || action.node->children.at(0)->result_name != sorting_key_columns[i])
-                {
+                if (!found_function)
                     current_direction = 0;
-                    break;
-                }
 
-                const auto & func = *action.node->function_base;
-                if (!func.hasInformationAboutMonotonicity())
-                {
-                    current_direction = 0;
-                    break;
-                }
+                if (!current_direction || (i > 0 && current_direction != read_direction))
+                    return false;
 
-                auto monotonicity = func.getMonotonicityForRange(*func.getArgumentTypes().at(0), {}, {});
-                if (!monotonicity.is_monotonic)
-                {
-                    current_direction = 0;
-                    break;
-                }
-                else if (!monotonicity.is_positive)
-                    current_direction *= -1;
+                if (i == 0)
+                    read_direction = current_direction;
+
+                return true;
             }
+        };
 
-            if (!found_function)
-                current_direction = 0;
+        const auto & actions = elements_actions[i]->getActions();
+        bool ok;
+        /// check if it's alias column
+        /// currently we only support alias column without any function wrapper
```

Comment:
So this is about `function(alias)` not `alias function(column)`, am I right?

Reply:
Yes.  `order by alias_column` can have this optimization,  but `order by function(alias)` can not.
This suits most cases.

Reply:
Agree, thank you